clang/lib/CConv/ConstraintVariables.cpp

```diff
@@ -123,6 +124,48 @@ PointerVariableConstraint::PointerVariableConstraint(DeclaratorDecl *D,
         PointerVariableConstraint(D->getType(), D, D->getName(),
                                   I, C) { }
 
+
+class TypedefLevelFinder : public RecursiveASTVisitor<TypedefLevelFinder> {
+  public:
+
+    static std::tuple<bool, int, std::string> find(const QualType &QT) {
+      TypedefLevelFinder TLF;
+      QualType tosearch;
+      if (auto TDT = dyn_cast<TypedefType>(QT))
+        tosearch = TDT->desugar();
+      else
+        tosearch = QT;
+      TLF.TraverseType(tosearch);
+      assert(IMPLIES(TLF.hastypedef, TLF.TDname != ""));
+      return std::make_tuple(TLF.hastypedef, TLF.typedeflevel, TLF.TDname);
+    }
+
+    bool VisitTypedefType(TypedefType *TT) {
+      hastypedef = true;
+      auto TDT = TT->getDecl();
+      TDname = TDT->getNameAsString();
+      return false;
+    }
+
+    bool VisitPointerType(PointerType *PT) {
+      typedeflevel++;
+      return true;
+    }
+
+    bool VisitArrayType(ArrayType *AT) {
+      typedeflevel++;
+      return true;
+    }
+
```

Comment:
Avoid extra spaces (here, and after the `private` block below).

@mattmccutchen-cci Can you suggest to Aaron how he might run `clang-format` to make the formatting consistent? That can be the formatting guide.

Reply:
> @mattmccutchen-cci Can you suggest to Aaron how he might run `clang-format` to make the formatting consistent?

Just run `clang-format -i clang/lib/3C/ConstraintVariables.cpp`.  However, I'm going to format this file as part of my bulk formatting to fix long lines, so formatting the whole file now in this PR might result in massive conflicts (I haven't thought through exactly what would happen).  Hopefully the PR can wait until I get the bulk formatting pushed (hopefully within a few hours).  Then the PR can be rebased onto (or can merge from) the new master with only the unavoidable conflicts, and the file in the PR can be reformatted after that.

> That can be the formatting guide.

Yes, I'll make a note to add this to the engineering handbook (just because that's what I'm comfortable updating now; all this stuff will need to move to the public contributor documentation later). :)

Reply:
Aaron: #333 is now merged, so you can update this PR to the new `master-post-microsoft` and then run `clang-format` on it whenever you are ready.  A `clang-tidy` bulk change is coming soon, but I don't think it will be nearly as disruptive, so I don't think you need to wait for it.